src/qutip_qoc/_rl.py

```diff
@@ -0,0 +1,313 @@
+        self.lbound = [b[0][0] for b in bounds]
+        self.ubound = [b[0][1] for b in bounds]
+
+        self._alg_kwargs = alg_kwargs
+
+        self._initial = objectives[0].initial
+        self._target = objectives[0].target
+        self.state = None
+        self.dim = self._initial.shape[0]
+
+        self._result = Result(
+            objectives = objectives,
+            time_interval = time_interval,
+            start_local_time = time.localtime(),    # initial optimization time
+            n_iters = 0,                            # Number of iterations(episodes) until convergence 
+            iter_seconds = [],                      # list containing the time taken for each iteration(episode) of the optimization
+            var_time = False,                       # Whether the optimization was performed with variable time
+        )
+
+        #for the reward
+        self._step_penalty = 1
+
+        # To check if it exceeds the maximum number of steps in an episode
+        self.current_step = 0
+
+        self.terminated = False
+        self.truncated = False
+        self.episode_info = []                # to contain some information from the latest episode
+
+        self._fid_err_targ = alg_kwargs["fid_err_targ"]
+
+        # inferred attributes
+        self._norm_fac = 1 / self._target.norm()
+
+        self.temp_actions = []                  # temporary list to save episode actions
+        self.actions = []                       # list of actions(lists) of the last episode
+
+        # integrator options
+        self._integrator_kwargs = integrator_kwargs
+        self._rtol = self._integrator_kwargs.get("rtol", 1e-5)
+        self._atol = self._integrator_kwargs.get("atol", 1e-5)
+
+        self.max_episode_time = time_interval.evo_time                  # maximum time for an episode
+        self.max_steps = time_interval.n_tslots                         # maximum number of steps in an episode
+        self.step_duration = time_interval.tslots[-1] / time_interval.n_tslots  # step duration for mesvole
+        self.max_episodes = alg_kwargs["max_iter"]                      # maximum number of episodes for training
+        self.total_timesteps = self.max_episodes * self.max_steps       # for learn() of gym
+        self.current_episode = 0                                        # To keep track of the current episode
+        
+        # Define action and observation spaces (Gym)
+        if self._initial.isket:
+            obs_shape = (2 * self.dim,)
+        else:   # for unitary operators 
+            obs_shape = (2 * self.dim * self.dim,)
+        self.action_space = spaces.Box(low=-1, high=1, shape=(len(self._Hc_lst[0]),), dtype=np.float32)     # Continuous action space from -1 to +1, as suggested from gym
+        self.observation_space = spaces.Box(low=-1, high=1, shape=obs_shape, dtype=np.float32)              # Observation space
+        
+    def update_solver(self): 
```

Comment:
I am afraid this method should not exist.
You can take a look at the _jopt.py or _goat.py files to see how you initialize a solver.
After initializing the solver once (in your __init__) there is no need to create the solver again.
It must take in the parameterized QobjEvo object and will update the amplitudes through the parameter args dictionary.